Add experimental modified (Planck/Rosseland) Fleck factor option.

RyanWollaeger opened the pull request:
## Background

* The modified Planck-Rosseland grey transport equation proposed by [Jiang (2021)](https://ui.adsabs.harvard.edu/abs/2021ApJS..253...49J/abstract) can be incorporated into IMC using a modified Fleck factor.
* The application is nominally limited to where this modified Fleck factor is <= 1.

## Description of Changes

+ Time-discretize Jiang grey Planck-Rosseland transport with IMC.
+ Use Planck opacity in Fleck factor; multiply by Planck/Rosseland.
+ Apply only when modified factor is <= 1.
+ Add Planck-only grey absorption option. 

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

RyanWollaeger committed:
Add experimental modified (Planck/Rosseland) Fleck factor option.

+ Time-discretize Jiang grey Planck-Rosseland transport with IMC.
+ Use Planck opacity in Fleck factor; multiply by Planck/Rosseland.
+ Apply only when modified factor is <= 1.

adamdempsey90 commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -253,6 +253,10 @@ Initialize_impl(ParameterInput *pin, EOS &eos,
 
+  // Experimental Fleck factor modification for grey Jiang Rossland-Planck transport
+  bool use_fj_factor = pin->GetOrAddBoolean(block_name, "use_fj_factor", false);
+  pkg->AddParam<>("use_fj_factor", use_fj_factor);
```
For the user facing parameter, maybe something more descriptive of what this is doing? Maybe two parameters called `use_planck`, `use_rosseland`? Then the code decides what to do based on the combination of those two.

RyanWollaeger replied:
That seems like a nice way to go. So if the user sets both true then they would get the modified Fleck factor, otherwise they get the usual grey formulation, with one or the other.

RyanWollaeger replied:
Added these - the change will require a slight change in artemis now, on the host side, to ensure the pre-computed opacities use the correct average type.

adamdempsey90 commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -448,6 +461,20 @@ TaskStatus UpdateDerivedTransportFieldsImpl(MeshData<Real> *md, const Real dt) {
+            const Real emisP = mopac.Emissivity(rho, temp, Planck);
+            Real fj = 1.0 / (1.0 + (4.0 * emisP / (rho * cv * temp)) * dt);
+            fj *= mopac.AbsorptionCoefficient(rho, temp, Planck) /
+                  mopac.AbsorptionCoefficient(rho, temp, Rosseland);
```
Need to guard against zero Rosseland?

RyanWollaeger replied:
Good catch here

RyanWollaeger replied:
The latest change avoids dividing by 0 Rosseland now. Also, the change brings in using Planck opacity for the DDMC leakage coefficients when `use_rosseland=false`. This is typically ill-advised since the leakage terms should theoretically use Rosseland opacity, but if Planck-only is wanted for absorption this is most consistent.

RyanWollaeger committed:
Incorporate suggestion from Adam Dempsey - permit using all-Planck.

+ Remove use_fj_factor paramter.
+ Add use_planck and use_rosseland input parameters.
+ Default to use_rosseland=true and use_planck=fals.
+ Use modified Fleck factor when use_planck=use_rosseland=true.
+ Use opacity type enum in singularity-opac Absorption/Emissivity.

RyanWollaeger committed:
+ Avoid dividing by a 0 Rosseland opacity (thanks A. Dempsey).
+ Use Planck opacity in DDMC leakage terms if use_rosseland=false.

RyanWollaeger committed:
Avoid known device lambda-capture issue with device if constexpr.

+ Pass Planck/Rosseland enum to [[maybe_unused]] device copy.

RyanWollaeger committed:
Merge branch 'develop' of github.com:lanl/jaybenne into fleck_jiang_factor

adamdempsey90 commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -446,15 +465,31 @@ TaskStatus UpdateDerivedTransportFieldsImpl(MeshData<Real> *md, const Real dt) {
+            // calculate modified fleck factor using Planck and Rosseland
+            const Real ross = mopac.AbsorptionCoefficient(rho, temp, Rosseland);
+            const Real plnk = mopac.AbsorptionCoefficient(rho, temp, Planck);
+            const Real f = vmesh(b, fj::fleck_factor(), k, j, i);
```
```suggestion
            const Real &f = vmesh(b, fj::fleck_factor(), k, j, i);
```

RyanWollaeger committed:
Make fleck_factor access const ref.

Co-authored-by: Adam M. Dempsey <adempsey@lanl.gov>

RyanWollaeger commented:
Thanks @pdmullen @adamdempsey90!
